File: ChallengeBook/Java/src/sample/PartialSum.java
package sample;

public class PartialSum {

	private int[] a;
	private int n, k;
	
	public static void main(String[] args) {
		new PartialSum().dfs(0, 0);
	}

	private boolean dfs(int i, int sum) {
		if (i == n) return sum == k;
		
		if (dfs(i + 1, sum)) return true;
		
		if (dfs(i + 1, sum + a[i])) return true;
		
		return false;
	}
}

File: ChallengeBook/CSharp/Original/PartialSum.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;

namespace ChallengeBook.Original {
	public class PartialSum {
		public bool Solve(int sum, IList<int> numbers) {
			return Dfs(0, sum, numbers);
		}

		private bool Dfs(int iNumbers, int remain, IList<int> numbers) {
			if (remain == 0) {
				return true;
			}
			if (iNumbers >= numbers.Count) {
				return false;
			}
			return Dfs(iNumbers + 1, remain - numbers[iNumbers], numbers) ||
				Dfs(iNumbers + 1, remain, numbers);
		}
	}
}
